### ISSUE TITLE
Ungated benchmark run hangs indefinitely right after Stage 3, blocking #30

### ISSUE THREAD

**Issue by ACFHarbinger** [ACFHarbinger/Anime-Stitch-Pipeline#49]
Blocks #30 (post-M1 ungated 97-run). Running the 5-case smoke check
(`just bench::asp-benchmark-ungated-verify`) hangs indefinitely — killed
after 30+ min on the first attempt, reproduced a second time hanging
again (this time on a different first test case, so not data-specific
to one test).

**Symptom:** log output stops dead right after Stage 3 completes and
Stage 4 (masking) begins:

    [Stitch]   Fitting BaSiC flat-field (spatial correction only)…
    [Stitch]   Vignette correction applied (k=0.0899).
    Loading weights: 100%|██████████| 447/447 [00:00<00:00, ...it/s]
    <nothing else, ever>

No error, no traceback, process stays alive (confirmed via `ps`), CPU
~110-218% (busy, multi-threaded), GPU utilization ~4% (`nvidia-smi`,
effectively idle). Reproduced twice, killed both times after 15-30+ min
with zero further log progress.

**What I ruled out** (isolated `BiRefNetWrapper` directly against real
corpus frames from `dump/asp_test09/`, outside the benchmark harness):
- Device selection is correct — `torch.cuda.is_available()` is `True`
  in the venv (torch 2.9.1+cu128), wrapper constructs with
  `device=cuda`.
- `w.load()` — 1.6s, fine (the "Loading weights: 447/447" tqdm line
  comes from `transformers/core_model_loading.py`'s generic
  weight-mapping loop during this call, not a separate model).
- `w.get_mask(single_frame)` — 0.51s.
- `w.get_mask_batch(22_frames)` — 5.35s.

So `BiRefNetWrapper` in isolation is fast and correctly GPU-resident.
The actual masking-loop caller is `_compute_fg_masks()`
(`backend/src/ingestion/masking.py:25-53`) — calls
`birefnet_wrapper.get_background_mask(img, dilate_px=FOREGROUND_DILATION,
erode_px=FOREGROUND_EROSION)` per-frame in a plain Python loop, same API
and same default args my isolated test used. I could not find an
obvious difference between my working isolated call and the hanging
in-pipeline call before running out of time on this.

**Separately, worth knowing (probably not the hang, but flag it):**
`run_stage.py`'s Stage 4 does
`from backend.src.models.wrappers.birefnet_wrapper import BiRefNetWrapper`
— the bare `backend.*` namespace, not ASP's collision-safe
`asp_backend.*` alias every other import in that file uses. When run via
the documented Image-Toolkit-venv+PYTHONPATH pattern (Image-Toolkit
listed first), this silently resolves to Image-Toolkit's own
`backend/src/models/wrappers/birefnet_wrapper.py`, not anything living
in ASP. Same for `BaSiCWrapper`/`LoFTRWrapper` imports nearby. This
looks like an intentional shared-dependency design (ASP has no local
copies of these three, unlike jamma/roma/efficient_loftr/aliked_lg which
it does have locally) rather than a migration bug, but it's fragile —
worth an explicit, named cross-repo import mechanism instead of relying
on PYTHONPATH ordering, and worth confirming Image-Toolkit's
`birefnet_wrapper.py` is actually the version ASP intends to depend on.

**Repro:**
```sh
cd submodules/ASP
just --justfile ../../justfile asp-just bench::asp-benchmark-ungated-verify
```

**Most useful next step:** a live py-spy/gdb stack trace while it's
hung — I don't have ptrace permission in this environment
(`py-spy dump` failed with Permission Denied, no passwordless sudo
available) to get one myself. Whoever picks this up with local/sudo
access should be able to nail the exact hung call in under a minute
with `py-spy dump --pid <pid>`.

Not GPU-starved (125GB RAM, RTX 3090 Ti with headroom, confirmed via
`free -h`/`nvidia-smi` during the hang) — this isn't a resource
contention issue, something is actually stuck.

**Comment by ACFHarbinger** [comment]
Landed `a45a96f`. Isolated BiRefNet was never the hang; the pipeline path mixes OpenCV OpenCL (default-on, NVIDIA 3090 Ti) with PyTorch CUDA after BaSiC, then per-frame `empty_cache`/`synchronize`.

Fix: disable OpenCL at bench start and `run()`; `_compute_fg_masks` uses `get_mask_batch`; flushed per-frame progress. 14 masking tests green. `#30` smoke is ready to retry.

**Comment by ACFHarbinger** [comment]
Re-ran the smoke check (\`just bench::asp-benchmark-ungated-verify\`) to verify. Good news: Gemini's progress logging + Grok's masking fix are both real and working — asp_test04 completed, and asp_test09's masking stage now shows clean progress (\`[Stitch] Generating BiRefNet foreground masks (22 frames)...\` → \`BiRefNet batch masks ready (22 frames).\`), no stall there this time.

But it hung again — a **second, different** stall point, right after masking, at another \`Loading weights: 100%|447/447\` line (a different model this time, most likely the LoFTR/JamMa/RoMa alignment/matching stage Gemini's diagnosis also mentioned: "\`_pairwise_match_with\` (28 LoFTR/JamMa image pairs on GPU)"). Same signature as before: GPU ~4% idle (\`nvidia-smi\`), process alive and CPU-busy, zero log progress for 12+ minutes before I killed it.

This matches Grok's root-cause theory (OpenCV OpenCL bound to the same GPU as PyTorch CUDA, contending/livelocking) continuing at the *next* stage that loads a model onto CUDA — if the OpenCL-disable fix was scoped to the masking pathway specifically rather than globally (e.g. at process start), the same class of stall would recur at the next GPU model load, which is exactly what I'm seeing.

Reopening — #30 is not actually unblocked yet. Whoever picks this up: check whether OpenCV OpenCL gets disabled once globally at bench import/process start vs. only around the masking call, and whether the alignment/matching stage's model-loading path has the same missing progress-logging Gemini added for masking.

**Comment by ACFHarbinger** [comment]
Follow-up, correcting my own earlier guess: checked the actual fix rather than re-running live again. The \`cv2.ocl.setUseOpenCL(False)\` call in \`run_stage.py\` is placed at the very top of \`run()\` — before *all* stages, not scoped to masking specifically — so my "OpenCL disable was too narrowly scoped" theory is wrong. Also ruled out the \`contextlib.suppress(Exception)\` silently swallowing a failed disable: tested directly, \`cv2.ocl.setUseOpenCL(False)\` works cleanly and \`useOpenCL()\` correctly reports \`False\` afterward.

So the second stall (right after masking, at the alignment/matching stage's own model-loading) has a different cause than OpenCL-vs-CUDA contention — OpenCL should already be off by that point. Likely something specific to how the LoFTR/JamMa/RoMa wrapper(s) load/manage GPU state in that stage, or it genuinely needs the same missing-progress-log treatment Gemini added for masking (i.e. maybe not a hang at all, same as Gemini's first read on masking) — I don't have a live stack trace to tell which, and don't have ptrace/sudo access here to get one. Leaving this for whoever has that next, with the OpenCL theory now ruled out so nobody re-chases it.

**Comment by ACFHarbinger** [comment]
Good news / bad news on 763a3fb's matching fix. Re-ran the smoke check live: masking and matching are both genuinely fixed now — asp_test04 completed fully (opencv_stitch.png + panorama written), asp_test08 completed fully, and asp_test09's matching cleared cleanly too (60/60 pairs, ~10s).

But after 90 minutes total, GPU dropped back to idle (4%) again — a **third** stall, this time right after test09's matching finished (so somewhere in bundle-adjustment/ECC refinement or compositing/rendering, whichever stage comes next for that case). No progress log, process alive, same signature as the first two. Killed it.

Given the pattern (three separate stages now: masking, matching, and whatever's next) each individually needed the same "add progress logging + stop calling cuda.synchronize()/empty_cache() unnecessarily per-item" treatment, it might be worth a sweep of the *whole* pipeline for this same anti-pattern (any loop that calls torch.cuda.synchronize()/empty_cache() per-iteration) rather than fixing stages one at a time as they're each individually discovered by a live run timing out.

Separately, unrelated but worth knowing: added HF_HUB_OFFLINE/TRANSFORMERS_OFFLINE=1 as defaults on the ungated bench recipes (3199add) — BiRefNet/ToonOut/EfficientLoFTR are all already cached locally, no reason for a multi-hour run to touch the network. Not related to this stall (this one's a compute stage, not a download), just landed while investigating.

**Comment by ACFHarbinger** [comment]
0d00a15 is real progress — verified live, asp_test04 and asp_test08 both completed their full pipelines in under 2 minutes total (down from 90+ min for less than that before). But asp_test09 (the exact case that stalled twice before) hung again, this time for 60+ seconds with zero progress right between \`Pair 60/60\` finishing and the very next line of code, \`[PostMatch] filtering N edges...\` (bench_anime_stitch.py:1487-ish) -- that print should be near-instant.

**Found the precise cause**: \`backend/src/models/wrappers/efficient_loftr_wrapper.py:76-81\`, \`EfficientLoFTRWrapper.offload()\`:
\`\`\`python
def offload(self) -> None:
    if self._model is not None:
        self._model.cpu()
        import torch
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
\`\`\`
This is called right after matching finishes (\`bench_anime_stitch.py\` ~line 1480, \`loftr.offload()\`). Same anti-pattern as the other two stalls, but it's an internal call *inside the wrapper method itself* -- 0d00a15 removed the *caller's* redundant \`torch.cuda.empty_cache()\` right after this same offload call, but not the one baked into \`offload()\` itself, so it's still there.

Killed the run, not re-verifying myself again -- this one's precise enough that whoever picks it up should be able to fix + verify in one pass without needing another live-debug session.

**Comment by ACFHarbinger** [comment]
Confirmed fully fixed. Ran the full 5-case smoke check live end-to-end after Codex's 8011135/0c47eba: completed in ~7.5 minutes total, exit code 0, 'All done. 5 datasets processed.', report + JSON artifacts written correctly. Watched asp_test09 specifically (the case that stalled 3 separate times across this investigation) go through masking -> matching -> spatial dedup -> phase detection -> edge filtering -> matcher unload -> connectivity check -> bundle adjustment -> a clean bounded SCANS fallback ('22 frames exceed the PANORAMA safety limit, using SCANS') instead of silently launching an expensive full-res OpenCV panorama stitch. All four distinct stall points found across this investigation (masking's per-frame empty_cache, matching's per-pair sync, the diagnostic resource-flush after every stage, and the unbounded PANORAMA fallback) are resolved.

Root cause chain, for the record:
1. Grok (a45a96f) — masking: OpenCV OpenCL contending with PyTorch CUDA + per-frame empty_cache/synchronize.
2. Codex (763a3fb) — matching: per-pair torch.cuda.synchronize()/empty_cache() serializing the fallback chain.
3. Codex (0d00a15) — diagnostic _log_resource()'s permanent empty_cache() after every stage transition.
4. Codex (8011135, 0c47eba) — the actual root cause of the last stall: failed affine validation silently launched OpenCV PANORAMA on full-resolution frames with no bound; now falls back directly to bounded SCANS (ASP_PANORAMA_MAX_FRAMES configurable), plus swept the same offload-flush anti-pattern across all matcher wrappers (aliked_lg, jamma, roma, efficient_loftr).

91 relevant ASP tests + 109 root wrapper tests green (verified independently), on top of Codex's own reported 134 + 38.

#30 is genuinely unblocked now.
